### PR TITLE
[improve][test] Refactor TestPulsarService to PulsarTestContext and add support for starting

### PR DESCRIPTION
**Pull request by lhotari**
### Motivation

This is a follow up on previous PRs (#19326, #19323)  to refactor Pulsar unit tests to reduce mocking which causes flakiness since Mockito isn't thread safe.

There are 2 types of Pulsar unit tests that use a PulsarService:
- Some Pulsar unit tests use a PulsarService that isn't started
- Some Pulsar unit tests start the PulsarService and use less mocking

With the changes in this PR, the test framework will be suitable for use across a broader set of Pulsar unit tests and we can get rid of more of the problematic usages of Mockito.

### Modifications

This PR refactors the previous TestPulsarService to PulsarTestContext which is more maintainable. 
It supports both modes of PulsarServices, startable and non-startable.
OwnerShipForCurrentServerTestBase is used as piloting the new mode. More tests will be converted after this PR has been merged.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/lhotari/pulsar/pull/138

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/namespace/OwnerShipForCurrentServerTestBase.java [hunk 1]
```diff
@@ -118,96 +86,22 @@ protected void startBroker() throws Exception {
...
 
-            PulsarService pulsar = spyWithClassAndConstructorArgs(PulsarService.class, conf);
-
-            setupBrokerMocks(pulsar);
-            pulsar.start();
+            PulsarTestContext.Builder testContextBuilder =
+                    PulsarTestContext.startableBuilder()
+                            .config(conf);
+            if (i > 0) {
+                testContextBuilder.reuseMockBookkeeperAndMetadataStores(pulsarTestContexts.get(0));
```
Comment (michaeljmarshall) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/namespace/OwnerShipForCurrentServerTestBase.java:93]: Why are we getting the first one here?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/testcontext/PulsarTestContext.java [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+import org.apache.zookeeper.MockZooKeeper;
+import org.apache.zookeeper.MockZooKeeperSession;
+import org.apache.zookeeper.data.ACL;
+import org.jetbrains.annotations.NotNull;
+
+@Slf4j
+@ToString
+@Getter
+@Builder(builderClassName = "Builder")
+public class PulsarTestContext implements AutoCloseable {
```
Comment (michaeljmarshall) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/testcontext/PulsarTestContext.java:74]: A Javadoc explaining how to use this class would be valuable. Also, are you able to point out what of this class is new? I see that it is pretty similar to the `TestPulsarService`, and I know your PR description indicates that it is an improvement on `TestPulsarService`.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/testcontext/NonStartableTestPulsarService.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+import org.apache.pulsar.common.naming.TopicName;
+import org.apache.pulsar.compaction.Compactor;
+import org.apache.pulsar.metadata.api.MetadataStore;
+import org.apache.pulsar.metadata.api.extended.MetadataStoreExtended;
+
+/**
+ * Subclass of PulsarService that is used for some tests.
+ * This was written as a replacement for the previous Mockito Spy over PulsarService solution which caused
+ * a flaky test issue https://github.com/apache/pulsar/issues/13620.
+ */
```
Comment (michaeljmarshall) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/testcontext/NonStartableTestPulsarService.java:54]: I see this was copied from `TestPulsarService`. Is there any new context we need in order to document how to use this class?

**Review by michaeljmarshall — COMMENTED** [review]
Overall, LGTM. I left some comments/questions.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/namespace/OwnerShipForCurrentServerTestBase.java (hunk 1 above)
Comment (lhotari) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/namespace/OwnerShipForCurrentServerTestBase.java:93]: PulsarTestContext holds the mock bookkeeper and metadata store (mockzookeeper) references. The "reuseMockBookkeeperAndMetadataStores" can be used to share the mock bookkeeper and zookeeper from another context.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/testcontext/PulsarTestContext.java (hunk 2 above)
Comment (lhotari) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/testcontext/PulsarTestContext.java:74]: I agree. I can improve the documentation in a follow-up PR that I have in progress.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/testcontext/NonStartableTestPulsarService.java (hunk 3 above)
Comment (lhotari) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/testcontext/NonStartableTestPulsarService.java:54]: I'll update the docs in a follow-up PR.

**Review by michaeljmarshall — APPROVED** [review]
LGTM. Thanks for addressing my questions.

**Pull request merged**